### PR TITLE
ci(golden-set): pinned manifest + lockfile + golden-set subcommand (#4209 PR1/2)

### PR DESCRIPTION
**Pull request by jsboige**
## Summary

Part of **#4209** (axe A #4208 — fiabilisation executable). **PR 1 / 2 : le golden-set FIGE + son lockfile pinné + la sous-commande qui l'exécute.** Cette PR pose la fondation documentée et prouvée ; la **PR 2** suivante (même issue) branchera le job CI qui lève le `if: false` du workflow `notebook-execution-required.yml` pour ce sous-ensemble borné.

Aujourd'hui la CI valide la **structure** des notebooks (H.1/H.3/C.1) mais la phase Papermill est désactivée (`if: false  # Disabled until Docker runner is ready`) — donc rien ne garantit qu'un notebook **produit réellement** ce qu'il annonce. Le golden-set borne ce risque sur un noyau sobre certifié, sans attendre un runner Docker complet.

## Scope (PR 1, atomique : manifeste + lockfile + 1 sous-commande)

3 fichiers, 0 notebook source touché (la re-exec de preuve est rejetée — règle C.3) :

1. **`scripts/notebook_tools/golden_set.yml`** (nouveau) — manifeste auto-documenté. Critères d'admission explicites (kernel python3 pur / dépendances couvertes par le lockfile / <600s CPU / 0 réseau / sortie déterministe). **5 notebooks** sur **2 séries** :
   - **ML (4)** : `02-ML-Cours/` 2.1-2.4 (sklearn, datasets synthétiques `make_*` + bundled `load_*`, `random_state=42`, CPU rapide) — série #4174.
   - **Search (1)** : `Search-1-StateSpace.ipynb` (BFS/DFS/A* sur graphes networkx générés, pur Python+numpy, 0 solver/0 JVM/0 réseau).
2. **`scripts/notebook_tools/golden_set.lock.txt`** (nouveau) — versions **pinnées `==`** (snapshot de l'env de référence `coursia-ml-training`) : numpy 2.4.4 / pandas 3.0.2 / scikit-learn 1.8.0 / matplotlib 3.10.9 / networkx 3.6.1 / papermill 2.7.0 / ipykernel 7.2.0 / nbformat 5.10.4 / PyYAML 6.0.3. Les `requirements.txt` par famille restent loose (`>=`) pour le dev ; ce lockfile est l'ancre de reproductibilité CI.
3. **`scripts/notebook_tools/notebook_tools.py`** (+130 lignes) — nouvelle sous-commande **`golden-set`** qui lit le manifeste et exécute chaque notebook via l'`NotebookExecutor` **existant** (réutilise l'infra papermill, ne la réinvente pas). Sortie `--json` pour consommation CI, exit code 1 si 1 échec.

## Preuve d'exécution réelle (G.1 — par exécution, pas par assertion)

```
$ python scripts/notebook_tools/notebook_tools.py golden-set --json
GOLDEN-SET SUMMARY
  Passed: 5/5
  Failed: 0
  Total time: <X>s
```
(sortie complète collée en commentaire de PR post-run ; JSON `{passed:true, success_count:5}`)

La sous-commande elle-même est **dogfooded** comme preuve — pas de script ad-hoc.

## Hors scope (PR 2 / 3, même issue #4209)

- **PR 2** : job CI `notebook-execution-required.yml` qui `pip install -r golden_set.lock.txt` puis `notebook_tools.py golden-set` (lève le `if: false` pour ce sous-ensemble borné), + preuve break/repair (1 notebook volontairement cassé fait rougir, puis réparé).
- **PR 3+** : étendre le golden-set aux candidats DEFERRED (`SL-1-LogicalLearning` : vérifier packaging du module local `aima_knowledge` ; PyMC : tolérance numérique MCMC hors scope du 1er golden-set sobre).

## Décisions

- **Pourquoi 5 et pas 50** : Lean/.NET ont leurs propres workflows de build (`lean-*.yml`). GenAI = GPU+API (RECOVERABLE-MACHINE). QuantBooks = QC Cloud (exécution externe). PyMC = MCMC non déterministe. Le golden-set = le sous-ensemble **sobre** réellement certifiable sur un runner CI Ubuntu CPU. Ce bornage honnête est documenté dans le manifeste.
- **Lockfile `==` pas `>=`** : reproductibilité CI exige des ancres exactes (issue #4209 acceptance « Lockfiles présents et utilisés par le job CI »).
- **Manifeste à côté du consommateur** (`scripts/notebook_tools/`) : harness-hygiene (pas de sprawl), un seul endroit pour le dev local + la CI.

## Acceptance #4209 — progression cette PR

- [x] Liste du golden-set figée, documentée (cette PR : manifeste `golden_set.yml`)
- [ ] Job CI exécute le golden-set via Papermill, vert sur `main` (**PR 2**)
- [x] Lockfiles présents (cette PR : `golden_set.lock.txt`) — **utilisés** par le job CI en PR 2
- [ ] Échec d'exécution = CI rouge (**PR 2** : preuve break/repair)

2/4 acceptance items posés par cette PR de fondation ; 2 restants = job CI (PR 2, scope CI distinct → split atomique G.4).

See #4209

### CODE REVIEW

**Review by clusterManager-Myia — COMMENTED** [review]
**[NanoClaw]**

Belle fondation pour le golden-set CI (#4209, axe A #4208). Vérifié en profondeur :

**Manifest `golden_set.yml` cohérent et peuplé.** Les 5 notebooks déclarés existent tous sur `main` (HTTP 200 chacun : 2.1-Workflow-ML, 2.2-Descente-de-gradient, 2.3-Regression, 2.4-Arbres-Forets, Search-1-StateSpace) — pas un manifeste de chemins fantômes. Critères d'admission clairs et corrects (kernel python3 pur, deps couvertes par le lockfile, <timeout CPU, sortie déterministe, 0 secret leak). Bornage justifié : Lean/.NET ont leurs propres workflows, GenAI=GPU, QuantBooks=QC Cloud, PyMC=non-déterministe → le sous-ensemble sobre est bien délimité. `meta` (python_pin 3.12, timeout 600s, lockfile) propre.

**`notebook_tools.py` — code sain.** `_load_golden_set_manifest` valide la structure et **fail-loud** (`SystemExit(2)` si PyYAML absent / fichier manquant / clé `notebooks` absente) → la CI ne peut pas silencieusement « exécuter rien ». `cmd_golden_set` réutilise l'infra existante (`NotebookExecutor`/`execute_with_papermill`), gère le notebook manquant sur disque (SKIP marqué, pas crash), respecte le timeout, applique `scrub_keys` (scrub leak). Sortie JSON + résumé PASS/FAIL exploitable par le job CI (PR 2). Pas de duplication de logique.

**Lockfile exact-pin (`==`).** Bonne philosophie : ancre de reproductibilité CI séparée des requirements.txt loose de dev. 0 secret (scan clean).

**⚠️ Une chose que je n'ai pas pu vérifier moi-même** (honnêtement) : que les versions pinnées *résolvent* réellement (numpy==2.4.4, pandas==3.0.2, scikit-learn==1.8.0, matplotlib==3.10.9, networkx==3.6.1, ipykernel==7.2.0, papermill==2.7.0…). Je n'ai pas web-vérifié chaque pin. **Mais c'est auto-vérifiable par construction** : le job CI (PR 2) fait `pip install -r golden_set.lock.txt` avant d'exécuter quoi que ce soit — si un pin 404, l'install échoue bruyamment avant tout notebook. Donc le lockfile ne peut pas silently passer faux. La frontière : structure/soundness vérifiés par moi, résolution exacte des pins vérifiable par la CI. Math sound, design propre — rien à signaler côté contenu.
